iroh-sync/src/ranger.rs

```diff
@@ -168,205 +164,73 @@ impl<K, V> MessagePart<K, V> {
-    }
-
-    /// Calculate the fingerprint of the given range.
-    fn get_fingerprint(&self, range: &Range<K>) -> Result<Fingerprint, Self::Error> {
-        let elements = self.get_range(range.clone())?;
-        let mut fp = Fingerprint::empty();
-        for el in elements {
-            let el = el?;
-            fp ^= el.0.as_fingerprint();
-        }
-
-        Ok(fp)
-    }
-
-    /// Insert the given key value pair.
-    fn put(&mut self, k: K, v: V) -> Result<(), Self::Error> {
-        self.data.insert(k, v);
-        Ok(())
-    }
-
-    type RangeIterator<'a> = SimpleRangeIterator<'a, K, V>
-        where K: 'a, V: 'a;
     /// Returns all items in the given range
-    fn get_range(&self, range: Range<K>) -> Result<Self::RangeIterator<'_>, Self::Error> {
-        // TODO: this is not very efficient, optimize depending on data structure
-        let iter = self.data.iter();
-
-        Ok(SimpleRangeIterator {
-            iter,
-            range: Some(range),
-        })
-    }
-
-    fn remove(&mut self, key: &K) -> Result<Option<V>, Self::Error> {
-        let res = self.data.remove(key);
-        Ok(res)
-    }
-
-    type AllIterator<'a> = SimpleRangeIterator<'a, K, V>
-    where K: 'a,
-          V: 'a;
-
-    fn all(&self) -> Result<Self::AllIterator<'_>, Self::Error> {
-        let iter = self.data.iter();
-
-        Ok(SimpleRangeIterator { iter, range: None })
-    }
-}
-
-#[derive(Debug)]
-pub struct SimpleRangeIterator<'a, K: 'a, V: 'a> {
-    iter: std::collections::btree_map::Iter<'a, K, V>,
-    range: Option<Range<K>>,
-}
-
-impl<'a, K, V> Iterator for SimpleRangeIterator<'a, K, V>
-where
-    K: RangeKey + Clone,
-    V: Clone,
-{
-    type Item = Result<(K, V), Infallible>;
+    fn get_range(&self, range: Range<E::Key>) -> Result<Self::RangeIterator<'_>, Self::Error>;
 
-    fn next(&mut self) -> Option<Self::Item> {
-        let mut next = self.iter.next()?;
+    /// Get all entries in the store
+    fn all(&self) -> Result<Self::RangeIterator<'_>, Self::Error>;
 
-        let filter = |x: &K| match &self.range {
-            None => true,
-            Some(ref range) => range.contains(x),
-        };
-
-        loop {
-            if filter(next.0) {
-                return Some(Ok((next.0.clone(), next.1.clone())));
-            }
-
-            next = self.iter.next()?;
-        }
-    }
+    /// Remove an entry from the store.
+    fn remove(&mut self, key: &E::Key) -> Result<Option<E>, Self::Error>;
 }
 
 #[derive(Debug)]
-pub struct Peer<K, V, S: Store<K, V> = SimpleStore<K, V>>
-where
-    K: RangeKey + Clone + Default + AsFingerprint,
-{
+pub struct Peer<E: RangeEntry, S: Store<E>> {
     store: S,
     /// Up to how many values to send immediately, before sending only a fingerprint.
     max_set_size: usize,
     /// `k` in the protocol, how many splits to generate. at least 2
     split_factor: usize,
 
-    _phantom: PhantomData<(K, V)>, // why???
+    _phantom: PhantomData<E>, // why???
```

Comment:
why why? E is not used directly, so you have to fake use using PhantomData.

Reply:
well, `Store` uses it, so I think it is dumb that this is needed..

Reply:
Added a comment